File: 416.partition-equal-subset-sum.cs
/*
 * @lc app=leetcode id=416 lang=csharp
 *
 * [416] Partition Equal Subset Sum
 */

// @lc code=start
public class Solution
{
    public bool CanPartition(int[] nums)
    {
        int expectedSum = 0;
        foreach (int num in nums)
            expectedSum += num;

        if (IsOddInteger(expectedSum))
            return false;
        expectedSum /= 2;

        bool[] dp = new bool[expectedSum + 1];
        dp[0] = true;

        foreach (int currentNumSelection in nums)
            for (int i = expectedSum; i > 0 && i - currentNumSelection >= 0 && !dp[expectedSum]; i--)
                dp[i] = dp[i] || dp[i - currentNumSelection];

        return dp[expectedSum];

    }

    public bool IsOddInteger(int value) => (value & 1) != 0;
}
// @lc code=end

File: 416.partition-equal-subset-sum.js
/*
 * @lc app=leetcode id=416 lang=javascript
 *
 * [416] Partition Equal Subset Sum
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canPartition = function (nums) {
  var sum = nums.reduce((a, b) => a + b);
  if (sum % 2 != 0) return false;
  var dp = new Array(sum / 2 + 1).fill(false);
  dp[0] = true; //base case

  for (let i = 1; i < nums.length; i++) {
    for (let j = sum / 2; j >= 0; j--) {
      if (j - nums[i - 1] >= 0) {
        dp[j] = dp[j - nums[i - 1]] || dp[j];
        if (j == sum / 2 && dp[j]) {
          return true;
        }
      }
    }
  }
  return false;
};
// @lc code=end
